ruby/sdk-phase-1.md

```diff
@@ -0,0 +1,439 @@
+      normal there, and in Python we only offered package-level methods because that's normal there.
+    * 💭 Why allow access to an instance? Some users want to pass the activity context to another thread.
+  * 💭 Why `Temporalio::Activity::Context` instead of just `Temporalio::Activity` or putting helpers on base class or
+    something? It is a common requirement in our SDKs to be able to access the activity context from the environment.
+    Also, it is a common requirement to be able to do that from utility methods outside of the explicit activity call.
+* Activity can choose its executor, default is `:default`.
+  * See worker section for more information on executors.
+* Cancellation is reported to the executor.
+  * For thread pool executor (the default), this raises a cancellation exception via `Thread.raise`.
+  * For fiber executor, this is done via `Fiber.raise`.
+  * A method on the activity context, `shield`, can be given a block that will hold off on cancelling until after it is
+    done. Nested `shield` calls only raise after the last one is complete.
+* Type-safe invocation from workflows is TBD.
+
+## Workers
+
+```ruby
+module Temporalio
+  class Worker
+    # Run multiple workers until signal reached, block optional
+    def self.run(*workers, shutdown_signals: [], &block)
+      ...
+    end
+
+    def initialize(
+      client,
+      task_queue:,
+      activities: [],
+      activity_executors: nil,
+      ...,
+    )
+      ...
+    end
+
+    # Run until shutdown called or block completed (which calls shutdown)
+    def run(&block)
+      ...
+    end
+
+    # Shutdown worker
+    def shutdown()
+      ...
+    end
+  end
+end
+```
+
+* Following the API of the current Temporal Ruby SDK, workers can be run individually, but we will also provide a class
+  method to run several at once.
+  * 💭 Why is something needed to run multiple workers for the user? Because in Ruby they don't have easy ways of
+    running multiple blocking things at once without external libraries or forcing threads on users.
+* Like other SDK methods, `run` is async-capable.
+* `activity_executors` can be a hash of `Temporalio::Worker::ActivityExecutor`s to run certain types of activities.
+  * Default is `{:default: Temporalio::Worker::ActivityExecutor::ThreadPool.new(max_concurrent_activities), :fiber: Temporalio::Worker::ActivityExecutor::Fiber.DEFAULT}`
+  * Activity can choose its executor, default is `:default`.
+  * Registration time will fail if activity references an executor that does not exist.
+  * Worker run will fail immediately before even starting if it is not run with a fiber scheduler present but any
+    activities reference a `Fiber` executor.
+    * 💭 Why wait until worker run to fail? Workers can be created outside of an async environment, it's just important
+      that they are run within an async environment if there are async activities.
```

Comment:
Are we also thinking about workers that have a combination of sync and async activities?

Reply:
Yes, this may be common (and is common in Python). This is just saying _if_ any are async we make sure we're in an async environment. Sync activities will always work.